### DISCUSSION TITLE
hb_ot_layout_lookup_get_glyph_alternates returns nothing

### DISCUSSION THREAD

**Discussion by totake0224** [harfbuzz/harfbuzz#4146] [Q&A] [answered]
Hello.
I'm trying to get alternates character from my font.but I couldn't get it by hb_ot_layout_lookup_get_glyph_alternates function.
Please help.
I confirm my font has 'aalt' table and I check adobe's Illustrator shows alternates selector at the font. 
my code is like this.
```
//create buffer and set mystring
 hb_buffer_t *buffer = hb_buffer_create();
 hb_buffer_add_utf8(buffer, _mystring_, -1, 0, -1);
//load font
 hb_blob_t* blob = hb_blob_create_from_file(_myfontfilepath_);
 hb_face_t* hb_face = hb_face_create(blob, 0);
//get glyphs
hb_glyph_info_t* info = hb_buffer_get_glyph_infos(buffer,&info_count);
//get alternates
 for(int i=0;i<info_count;i++){
      hb_codepoint_t* alt_glyphs;
      unsigned int  count=10;
      alt_glyphs=malloc(sizeof(hb_codepoint_t)*count+1);
      unsigned int  number=hb_ot_layout_lookup_get_glyph_alternates(hb_face,lookup_index,info[i].codepoint,0,&count,alt_glyphs);
}
```
like this. hb_ot_layout_lookup_get_glyph_alternates return alway 0.
I got lookup_index by using
	hb_ot_layout_table_get_feature_tags
        hb_ot_layout_feature_get_lookups

Thank you.

**Comment by behdad** [comment]
> hb_ot_layout_table_get_feature_tags
> hb_ot_layout_feature_get_lookups

Show us that part as well. It must be an issue there then.

**Comment by totake0224** [comment]
Thank you for your response.
I found my mistake in my code that collect lookup index.
I misunderstood feature index is "hb_ot_layout_table_get_feature_tags" return array's index. but it's wrong.
so I changed it to "hb_ot_layout_collect_lookups"
but I have still problem. I couldn't understand how to get value from hb_set_t.

  ```
hb_set_t*  indexes=hb_set_create();
  unsigned int feats[2];
  feats[0]=HB_TAG('a','a','l','t');
  feats[1]=HB_TAG_NONE;
  hb_ot_layout_collect_lookups(hb_face,HB_OT_TAG_GSUB,NULL,NULL,feats,indexes);
  unsigned int test = hb_set_get_population(indexes);
```

I think "hb_set_get_population"  returns collect value of my font.
but I couldn't found function to get value from hb_set_t from "https://harfbuzz.github.io/harfbuzz-hb-set.html"
I only found "hb_set_get_user_data ()" function but it needs "key" that I don't know.
I feel this question is very basic and it because my ignorance. but I couldn't found it by myself.
please give me some advice.
Thank you.

**Comment by totake0224** [comment, marked as answer]
I think I found a function that get value from hb_set_t. (thanks to 
 https://android.googlesource.com/platform/external/freetype/+/fb6b5b1/src/autofit/hbshim.c)
now, my code is below 

```
unsigned int get_aalt_lookup_index(hb_face_t *hb_face,unsigned int *n_lookup,unsigned int* lookups){
	hb_set_t*  indexes=hb_set_create();
	unsigned int feats[2];
	feats[0]=HB_TAG('a','a','l','t');
	feats[1]=HB_TAG_NONE;
	hb_ot_layout_collect_lookups(hb_face,HB_OT_TAG_GSUB,NULL,NULL,feats,indexes);

	unsigned int n=0;
	hb_codepoint_t idx;
	 for ( idx = HB_SET_VALUE_INVALID; hb_set_next(indexes, &idx ); ){
		if(*n_lookup>n){
			lookups[n]=idx;
		}
		n++;
	 }
	 if(*n_lookup>n){
		*n_lookup=n;
	 }
	return n;
}
```

I got lookup index [0] from this code. but I got same result from hb_ot_layout_lookup_get_glyph_alternates.
I wonder what means hb_ot_layout_lookup_get_glyph_alternates seconde variable [a glyph id] is.
In my recognition it's unicode character code. Is it collect ?

**Reply by behdad** [reply]
> I wonder what means hb_ot_layout_lookup_get_glyph_alternates seconde variable [a glyph id] is.
> In my recognition it's unicode character code. Is it collect ?

That is the problem indeed. A glyph-id is the output of `hb_shape()`. Unicode character is the *input* to `hb_shape()`.

**Reply by totake0224** [reply]
Thank you for your advice. I add hb_shape() before hb_buffer_get_glyph_infos() like this.
```
	hb_blob_t* blob = hb_blob_create_from_file(var);
	hb_face_t* hb_face = hb_face_create(blob, 0);
+	hb_font_t * hb_font = hb_font_create (hb_face);
+	hb_buffer_set_direction(buffer, HB_DIRECTION_LTR);
+	hb_shape(hb_font, buffer,NULL, 0);
```
but I got still nothing from hb_ot_layout_lookup_get_glyph_alternates() (this change only hb_ot_layout_lookup_would_substitute return value 0 to 1.)
Now I wonder , I want alternate chars that defined at 'aalt' table.(like attached picture)
hb_ot_layout_lookup_get_glyph_alternates is the function for that purpose ?

this is Fontforge table difinition of my font. 
![fontforge table difinition](https://user-images.githubusercontent.com/42058913/223333622-5e4059c7-bb92-459d-86b9-d802f54a4bcb.png)

**Reply by behdad** [reply]
Yes, looks correct. What font and text are you using?

**Reply by totake0224** [reply]
I'm using font published by a Japanese General Incorporated Association.
https://moji.or.jp/mojikiban/font/
there font has no aalt table. so I added it by myself. 
The licence allowed me attach it to this forum (I think)
but they gather kanji characters as many as they could. so that font size over 30Mb.  so I couldn't attach this font.
I will make small size font to reproduce my situation.

In addtion to that,
I test hb_ot_layout_lookup_get_glyph_alternates at another commercial font that have aalt table from origin but same result.

**Reply by totake0224** [reply]
I made very small font and did same thing.
I think I could get lookup index.and also I could get glyphs by [hb_ot_layout_lookup_collect_glyphs]
I confirm input and output hb_set_t of  above function. and it's number is same to aalt lookup table of the font.
but I could not get glyphs from [hb_ot_layout_lookup_get_glyph_alternates]
I'm confusing. 

this font I set aalt table for character 'a' 'b' 'c' 'd' 'e'. so I think if Glyph ID of 'a' given to [hb_ot_layout_lookup_get_glyph_alternates] then it will return 1. 
[MC_Sylp00-subst-aalt.zip](https://github.com/harfbuzz/harfbuzz/files/10915615/MC_Sylp00-subst-aalt.zip)

**Reply by behdad** [reply]
The problem seems to be that, at least in your test font, the lookup is of type SingleSubstFormat1, instead of AlternateSubstFormat1.

**Reply by behdad** [reply]
@khaledhosny Should we make that work nonetheless?

**Reply by behdad** [reply]
> The problem seems to be that, at least in your test font, the lookup is of type SingleSubstFormat1, instead of AlternateSubstFormat1.

Looks like at least AdobeVFPrototype also uses SingleSubst for some of its characters in aalt. It has two lookups, one SingleSubst, one AlternateSubst (for those with more than one substitution).

**Reply by behdad** [reply]
Okay I implemented that.

Please try again with main branch.

**Reply by behdad** [reply]
I also added code to show how to use this API: https://github.com/harfbuzz/harfbuzz/blob/main/src/test-gsub-get-alternates.cc

**Reply by totake0224** [reply]
Thank you.
I didn't get what is SingleSubstFormat1 and AlternateSubstFormat1 yet.
But at least my understanding of [hb_ot_layout_lookup_get_glyph_alternates] is collect now.I'm relieved.
I will search what is these Format is. and how to change it at Fontforge.if I found it. I will report it.

**Reply by behdad** [reply]
If you use HarfBuzz from main branch it should work with your existing font.

**Reply by totake0224** [reply]
I found how to set aalt table at Fontforge to make hb_ot_layout_lookup_get_glyph_alternates work.
Selection of table type is a key. I attach a picture. (I'm sorry it Japanese Fontforge. but everyone can guess which table type.)
Thank you.Mr.behdad.
I'm very much appreciate your help.
and I will very grad if someone make SingleSubstFormat1 and AlternateSubstFormat1 both work. because I have Japanese popular commercial font that hb_ot_layout_lookup_get_glyph_alternates doesn't work.(I'm sure Adobe's Illustrator can handle these font alternates.)

![FontForge_aalt_table](https://user-images.githubusercontent.com/42058913/223623051-3670c3f0-f222-4162-a100-355f5b164241.png)

**Reply by totake0224** [reply]
I have the utmost gratitude. Mr behdad.
I build latest harfbuzz from source and I could confirm both SingleSubstFormat1 and AlternateSubstFormat1 works.
hb_ot_layout_lookup_get_glyph_alternates return non-zero value at My commercial font also.

**Reply by behdad** [reply]
You're welcome. Thanks for reporting.

**Comment by behdad** [comment]
Please paste your full code (or as a github gist) so we can run it and debug it.

**Comment by totake0224** [comment]
Thank you. below is my code.
it's for PHP module so,there is PHP specific macros.
I explain each macro what is it.
```
unsigned int get_aalt_lookup_index(hb_face_t *hb_face,unsigned int *n_lookup,unsigned int* lookups){
	hb_set_t*  indexes=hb_set_create();
	unsigned int feats[2];
	feats[0]=HB_TAG('a','a','l','t');
	feats[1]=HB_TAG_NONE;
	hb_ot_layout_collect_lookups(hb_face,HB_OT_TAG_GSUB,NULL,NULL,feats,indexes);

	unsigned int n=0;
	hb_codepoint_t idx;
	 for ( idx = HB_SET_VALUE_INVALID; hb_set_next(indexes, &idx ); ){
		if(*n_lookup>n){
			lookups[n]=idx;
		}
		n++;
	 }
	 if(*n_lookup>n){
		*n_lookup=n;
	 }
	return n;
}

//-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
//main function
//this macro is entry point of PHP.
PHP_FUNCTION(harfbuzz_php_variants)
{
	char *var = "\0";
	size_t var_len = sizeof("World") - 1;
	char *var2 = "\0";
	size_t var2_len = sizeof("World") - 1;
	zend_long var3 = 1;  // it seems int64.
	zend_long var4 = 0;

//PHP macros to get input variables.
//this PHP function get 4 parameter but at this point I use only 2 parameter(var as fontfile path var2 as characters which I want to check alternatives)
	ZEND_PARSE_PARAMETERS_START(0, 3)
		Z_PARAM_OPTIONAL
		Z_PARAM_STRING(var, var_len)// getting input var from php  
		Z_PARAM_STRING(var2, var2_len)
		Z_PARAM_LONG(var3)
		Z_PARAM_LONG(var4)
	ZEND_PARSE_PARAMETERS_END();

	hb_buffer_t *buffer = hb_buffer_create();
	hb_buffer_add_utf8(buffer, var2, -1, 0, -1); //var2 is characters

	hb_blob_t* blob = hb_blob_create_from_file(var);//var is fontfile path
	hb_face_t* hb_face = hb_face_create(blob, 0);
	
	unsigned int info_count;
	hb_glyph_info_t* info = hb_buffer_get_glyph_infos(buffer,&info_count);

//below is PHP macros that return arrays to caller. so please think it as printf or something
//array_init
//add_assoc_***
//add_index_***

	array_init(return_value);
	unsigned int v1,v2,v3;
	hb_version (&v1,&v2,&v3);//my version of harfbuzz is 7.0.1. I build this from source.
	char version[100];
	sprintf(version,"%d.%d.%d",v1,v2,v3);
	add_assoc_string(return_value,"hb_version",version);
	unsigned int n_lookup=100;
	unsigned int* lookups=malloc(sizeof(unsigned int)* (n_lookup+1));
	get_aalt_lookup_index(hb_face,&n_lookup,lookups);
	zval zarray1;
	array_init(&zarray1);
	for(int i = 0 ; i < n_lookup; i++){
		add_index_long(&zarray1,i,lookups[i]);
	}
	add_assoc_zval(return_value,"lookups",&zarray1);

	for(int i = 0 ; i < info_count; i++){
		zval zarray;
		array_init(&zarray);
		add_assoc_long(&zarray,"unicode",info[i].codepoint);
		hb_bool_t result = hb_ot_layout_lookup_would_substitute (hb_face,lookups[0],&info[i].codepoint,1,0);//my font has only one aalt lookup. I wil fix this after I could get alternates
		add_assoc_long(&zarray,"would_substitute",result);
		add_assoc_long(&zarray,"has_subst",hb_ot_layout_has_substitution(hb_face));

		hb_codepoint_t* alt_glyphs;
		unsigned int  count=10;
		alt_glyphs=malloc(sizeof(hb_codepoint_t)*count+1);
		unsigned int number=0;
		number=hb_ot_layout_lookup_get_glyph_alternates(hb_face,lookups[0],info[i].codepoint,0,&count,alt_glyphs);//my font has only one aalt lookup. I wil fix this after I could get alternates
		add_assoc_long(&zarray,"lookup_index",lookups[0]);
		add_assoc_long(&zarray,"num_alternates",number);
		free(alt_glyphs);
		add_index_zval(return_value,i,&zarray);
	}
	free(lookups);
}

```
